src/python/gudhi/barycenter.py

```diff
@@ -0,0 +1,227 @@
+                              # that is G[j, i] = k <=> y_j is matched to
+                              # x_k in the diagram i-th diagram X[i]
+        updated_points = np.zeros((K, 2))  # will store the new positions of
+                                           # the points of Y.
+                                           # If points disappear, there thrown
+                                           # on [0,0] by default.
+        new_created_points = []  # will store potential new points.
+
+        # Step 1 : compute optimal matching (Y, X_i) for each X_i
+        #          and create new points in Y if needed
+        for i in range(m):
+            indices = _optimal_matching(Y, X[i])
+            for y_j, x_i_j in indices:
+                if y_j < K:  # we matched an off diagonal point to x_i_j...
+                    if x_i_j < nb_off_diag[i]:  # ...which is also an off-diagonal point
+                        G[y_j, i] = x_i_j
+                    else:  # ...which is a diagonal point
+                        G[y_j, i] = -1  # -1 stands for the diagonal (mask)
+                else:  # We matched a diagonal point to x_i_j...
+                    if x_i_j < nb_off_diag[i]:  # which is a off-diag point ! so we need to create a new point in Y
+                        new_y = _mean(np.array([X[i][x_i_j]]), m)  # Average this point with (m-1) copies of Delta
+                        new_created_points.append(new_y)
+
+        # Step 2 : Update current point position thanks to the groupings computed
+
+        to_delete = []
+        for j in range(K):
+            matched_points = [X[i][G[j, i]] for i in range(m) if G[j, i] > -1]
+            new_y_j = _mean(matched_points, m)
+            if not np.array_equal(new_y_j, np.array([0,0])):
+                updated_points[j] = new_y_j 
+            else: # this points is no longer of any use.
+                to_delete.append(j)
+        # we remove the point to be deleted now.
+        updated_points = np.delete(updated_points, to_delete, axis=0)  # cannot be done in-place.
+
+
+        if new_created_points: # we cannot converge if there have been new created points.
+            Y = np.concatenate((updated_points, new_created_points))
+        else:
+            # Step 3 : we check convergence
+            if np.array_equal(updated_points, Y):
+                converged = True 
+            Y = updated_points
+
+
+    if verbose:
+        groupings = []
+        energy = 0
+        log = {}
+        n_y = len(Y)
+        for i in range(m):
+            edges, cost = _optimal_matching(Y, X[i], withcost=True)
+            print(edges)
+            groupings.append([x_i_j for (y_j, x_i_j) in enumerate(edges) if y_j < n_y])
+            energy += cost
+            log["groupings"] = groupings
+        energy = energy/m
+        log["energy"] = energy
+
+        return Y, log
```

Comment:
This use case is not tested. Do you have any way to test it ?
A test of verbose mode would be useful. Maybe ok if you relaunch `test_lagrangian_barycenter` in verbose mode ?

Reply:
Tested now